### PR TITLE
feat(autox): support leaderboard column reordering + column presets for autorag

### PR DESCRIPTION
**Pull request by daniduong**
https://redhat.atlassian.net/browse/RHOAIENG-73378

## Description

Adds a new **Manage Columns** modal to the AutoRAG and AutoML leaderboard tables with drag-and-drop column reordering, visibility toggling, and quick-select presets.

### Changes

**New `ManageColumnsModal` component** (both packages):
- Drag-and-drop column reordering via PF's `DragDropSort` + `DataList`
- Checkbox per column to toggle visibility — all columns are toggleable (no locked columns)
- `BulkSelect` dropdown for select all / deselect all
- "Reset to default" button restores original column order and default visibility
- Quick Select section with "Organize by" dropdown for presets (AutoRAG only; hidden for AutoML)
- Preset auto-detection: dropdown label updates when current selection matches a preset (e.g. checking all columns shows "Full configuration")
- Presets and reset also restore default column order, not just visibility
- Save button disabled when no changes are made or zero columns are selected
- Cancel discards dirty state — reopening the modal always shows the last saved state

**AutoRAG leaderboard:**
- Default visible columns: Rank, Pattern name, Model name, Optimized metric (4 of 15)
- Three presets: "Optimization metrics", "Optimization metrics and chunking", "Full configuration"
- Column ordering follows RAG pipeline: Identity → Metrics → Chunking → Retrieval (number of chunks last)
- Fixed `onSaveNotebook` to pass `entry.patternKey` (S3 directory key) instead of `entry.pattern` (display name)

**AutoML leaderboard:**
- Default visible columns: Rank, Model name, Optimized metric (3 of N)
- Presets hidden until AutoML has meaningful preset groups

**Table changes (both):**
- Removed all left-sticky column positioning (`isStickyColumn`, `stickyMinWidth`, `stickyLeftOffset`) — only the Actions column remains sticky on the right
- Unified column rendering loop replaces separate hardcoded column blocks + dynamic `.map()`
- Toolbar shows "X/Y columns selected" count next to "Manage columns" button (right-aligned)
- Leaderboard wrapped in a PF `Card` with "Results" title
- Sort state consolidated into a single `activeSort` object (`{ id, direction }`)
- Sort fallback uses first visible column instead of hardcoding 'rank'

**Topology section (both):**
- "Experiment pipeline" title overlay spacing adjusted from `--spacer--md` (16px) to `--spacer--lg` (24px) to match Card padding

**Dependency changes (both):**
- Added `@patternfly/react-drag-drop` as explicit dependency in `package.json` and Module Federation shared singleton (was transitive-only)

### Key implementation decisions

- **Whitelist visibility** (`visibleColumnIds` Set) instead of blacklist (`hiddenColumnIds`) — new columns from API data loading are hidden by default
- **`defaultColumns` prop** on the modal — "Reset to default" uses this instead of `appliedColumns`, so it restores the true default state regardless of what was previously saved
- **Derived preset detection** via `useMemo` instead of tracking selected preset in state — auto-syncs when individual checkboxes change
- **`isOpen`-gated resync** — the modal's `useEffect` resyncs `currentColumns` from `appliedColumns` only when `isOpen` becomes true, preventing stale dirty state after cancel
- **Single `activeSort` object** instead of separate `activeSortId` + `activeSortDirection` states — avoids calling a setter inside another setter's updater
- **`aria-labelledby`** on column checkboxes points to a proper `id` attribute (was incorrectly referencing `data-ouia-component-id`)

## How Has This Been Tested?

- Type-checking passes (`npx tsc --noEmit`) for both autorag and automl frontends
- ESLint passes with zero warnings for all changed files
- All existing tests updated and passing (1156 autorag + 1207 automl)
- 14 new AutoRAG tests + 9 new AutoML tests covering edge cases

## Test Impact

**AutoRAG** (`AutoragLeaderboard.spec.tsx`): 79 tests (was 65)
- Updated 13 existing tests to use `showAllColumns()` helper (default visibility changed)
- Updated manage columns assertions (columns no longer `isUntoggleable`)
- Added mock for `@patternfly/react-drag-drop` to avoid `createPortal` jsdom issues
- New tests:
  - Default visibility (only 4 columns)
  - Column count display in toolbar
  - Reset to default restores original order and visibility
  - Organize-by presets render and apply correctly
  - Full configuration preset auto-detected when all columns checked
  - Sort resets when sorted column is hidden
  - No left-sticky columns on headers
  - Persisted column order after save
  - Reset restores column order (not just visibility)
  - Cancel discards dirty state on reopen
  - Save button disabled when no changes made
  - Save button enabled after toggling a column
  - Save button disabled when all columns unchecked

**AutoML** (`AutomlLeaderboard.spec.tsx`): 65 tests (was 56)
- Same pattern of updates as AutoRAG
- New tests: default visibility, column count, reset to default, persisted column order, reset restores order, cancel discards dirty state, save button disabled/enabled states

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

If you have UI changes:
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

## Summary by CodeRabbit

* **New Features**
  * Added a column management experience for results tables, including show/hide controls, drag-and-drop ordering, and preset views.
  * Updated leaderboard results to show a cleaner toolbar with selected-column counts and a “Manage columns” action.
* **Bug Fixes**
  * Improved table sorting so hidden columns no longer keep outdated sort state.
  * Refined loading and results layouts for a more consistent card-based presentation.
* **Style**
  * Adjusted overlay spacing in run results screens.

### CODE REVIEW

**Comment by daniduong** [comment]
@coderabbitai review

**Comment by daniduong** [comment]
@coderabbitai review

File: packages/autorag/frontend/src/app/components/run-results/AutoragLeaderboard.tsx [hunk 1]
```diff
@@ -390,85 +395,89 @@ function AutoragLeaderboard({
...
-        return 'rank';
-      }
-      return currentId;
-    });
+    // Reset sort to the first visible column if the currently sorted column is being hidden
+    const fallbackSortId = newColumns.find((col) => col.isShown)?.key ?? 'rank';
+    setActiveSort((prev) =>
+      newVisibleIds.has(prev.id) ? prev : { id: fallbackSortId, direction: 'asc' },
+    );
   }, []);
```
[the remark this answers is not in the document]
Comment (daniduong) [reply packages/autorag/frontend/src/app/components/run-results/AutoragLeaderboard.tsx:459]: Theoretically valid but practically a non-issue. patterns and optimizedMetric come from the pipeline run context and don't change for a given results page — they're fetched once from S3 and cached. There's no scenario where a metric column disappears while the user is looking at the leaderboard.

Adding an effect that runs on every sortableColumnIds change also risks unwanted sort resets during normal column toggling (which already has its own reset logic in handleApplyColumns). It would create two competing reset paths.

**Comment by daniduong** [comment]
@coderabbitai review

**Comment by daniduong** [comment]
@coderabbitai review

**Comment by nickmazzi** [comment]
Thanks @daniduong! Fixed merge conflict and pushed up.

# AI assisted review

My review skill did not find any concerns other than duplicated components (which cant be addressed right now).

From the [PR review toolkit ](https://github.com/anthropics/claude-code/tree/main/plugins/pr-review-toolkit)(I removed irrelevant items)

| # | Severity | Issue | Location |
|---|----------|-------|----------|
| 2 | Critical | throw inside setState updater crashes component tree on unexpected drag-drop input | ManageColumnsModal.tsx:124 (both) |
| 4 | Important | Save/Cancel buttons in ModalBody instead of ModalFooter (PF convention) | ManageColumnsModal.tsx:240-260 (both) |
| 5 | Important | Render helpers (getHeaderTestId, getCellTestId) defined inside component body | AutomlLeaderboard.tsx, AutoragLeaderboard.tsx |
| 6 | Important | Toolbar hasNoPadding — PF v5 prop, verify it works in PF v6 | Both leaderboards |
| 8 | Suggestion | handleDrop reorder logic never exercised (drag-drop fully mocked) | Tests (both) |
| 9 | Suggestion | No test for reorder-only hasChanges (only visibility changes tested) | Tests (both) |
| 10 | Suggestion | Automl missing "reset sort when sorted column hidden" test | AutomlLeaderboard.spec.tsx |


# Test Validation

## autorag

### Overall Summary

| Suite      | Result |
|------------|--------|
| Frontend   | ✅     |
| BFF        | ✅     |
| Contract   | ✅     |
| BFF Lint   | ✅     |
| Lint       | ✅     |
| Type Check | ✅     |

## automl

### Overall Summary

| Suite      | Result |
|------------|--------|
| Frontend   | ✅     |
| BFF        | ✅     |
| Contract   | ✅     |
| BFF Lint   | ✅     |
| Lint       | ✅     |
| Type Check | ✅     |

# Manual validation

Looks good in both components. Tested reordering and the presets in autorag.

**Comment by daniduong** [comment]
| # | Severity | Finding | Verdict |
| --- | --- | --- | --- |
| 2 | Critical | `throw` inside setState updater crashes component tree | Fixed — replaced with silent filter via `reduce`. |
| 4 | Important | Save/Cancel buttons in ModalBody instead of ModalFooter | Fixed — moved to `ModalFooter`. |
| 5 | Important | Render helpers defined inside component body | Skipped — plain functions returning strings, not components. No perf impact. |
| 6 | Important | `Toolbar hasNoPadding` is a PF v5 prop | Skipped — prop exists in PF v6 type defs and implementation. |
| 8 | Suggestion | `handleDrop` reorder logic never exercised in tests | True but unavoidable — drag-drop can't be simulated in jsdom. |
| 9 | Suggestion | No test for reorder-only `hasChanges` | Fixed — added `ManageColumnsModal.spec.tsx` for both packages. |
| 10 | Suggestion | Automl missing "reset sort when sorted column hidden" test | Fixed — added test to `AutomlLeaderboard.spec.tsx`. |

**Comment by nickmazzi** [comment]
Thanks for updating @daniduong I forgot about this one.
Just pulled and tested and looks good still. 

I realized a new metric is added in my PR for https://github.com/opendatahub-io/odh-dashboard/pull/8500 - `overall_score`

After this PR is merged I will merge main into mine and ensure overall_score is added to the optimization metrics column preset

/lgtm
/approve
